src/algebra/module/projective.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2021 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard
+-/
+
+import algebra.module.basic
+import linear_algebra.finsupp
+import linear_algebra.basis
+
+/-!
+
+# Projective modules
+
+This file contains a definition of a projective module, the proof that
+our definition is equivalent to a lifting property, and the
+proof that all free modules are projective.
+
+## Main definitions
+
+Let `R` be a ring and let `M` be an `R`-module.
+
+* `is_projective R M` : the proposition saying that `M` is a projective `R`-module.
+
+## Main theorems
+
+* `lifting_property` : a map from a projective module can be lifted along
+  a surjection.
+
+* `of_lifting_property` : If for all R-module surjections `A →ₗ B`, all
+  maps `M →ₗ B` lift to `M →ₗ A`, then `M` is projective.
+
+* `of_free` : Free modules are projective.
+
+## Implementation notes
+
+The actual definition of projective we use is that the natural R-module map
+from the free R-module on the type M down to M splits. This is more convenient
+than certain other definitions which involve quantifying over universes,
+and also universe-polymorphic (the ring and module can be in different universes).
+
+`of_lifting_property` is not universe polymorphic.
+
+## References
+
+https://en.wikipedia.org/wiki/Projective_module
+
+## Tags
+
+projective module
+
+-/
+
+universes u v
+
+/-- An R-module is projective if it is a direct summand of a free module.
+There are several other equivalent definitions. -/
+def is_projective
+  (R : Type u) [ring R] (M : Type v) [add_comm_group M] [module R M] : Prop :=
+∃ s : M →ₗ[R] (M →₀ R), ∀ m, finsupp.total M M R id (s m) = m
```

Comment:
The following is defeq but comes with a little more API:
```suggestion
∃ s : M →ₗ[R] (M →₀ R), function.right_inverse s (finsupp.total M M R id)
```
as then when you do `cases h with s hs`, you now have `hf.surjective` which states that `(finsupp.total M M R id)` is surjective.

Reply:
`hf.surjective` is about a different function `f`. I do not see what benefit this change brings in practice other than making the definition even harder to read. Is the point that this might not benefit this particular PR but might be more useful down the line?

Reply:
Conversely, I don't feel strongly about the suggestion; if `surjective (finsupp.total M M R id)` isn't remotely interesting then there's no point obfuscating things to make it easy to get. Is `function.injective s` any more interesting?

Reply:
If `surjective (finsupp.total M M R id)` is interesting then it can just be proved independently -- it's nothing to do with projective modules. However injectivity of `s` is something which I can believe is useful, so I'll go for the `left_inverse` option.

Reply:
I would now resolve this issue but I'm still confused by something -- `right_inverse f g` and `left_inverse g f` are synonymous (and presumably defeq) so why would injectivity of `s` be something which is easier to get from one than the other?